# **Workshop "Data Ingestion with dlt": Homework**


---

## **Dataset & API**

We’ll use **NYC Taxi data** via the same custom API from the workshop:

🔹 **Base API URL:**  
```
https://us-central1-dlthub-analytics.cloudfunctions.net/data_engineering_zoomcamp_api
```
🔹 **Data format:** Paginated JSON (1,000 records per page).  
🔹 **API Pagination:** Stop when an empty page is returned.  

## **Question 1: dlt Version**

1. **Install dlt**:

In [1]:
 !pip install dlt[duckdb]

  Using cached setuptools-75.8.0-py3-none-any.whl.metadata (6.7 kB)
  Using cached pycparser-2.22-py3-none-any.whl.metadata (943 bytes)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.2/20.2 MB 11.1 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 759.5/759.5 kB 11.5 MB/s eta 0:00:00
Using cached setuptools-75.8.0-py3-none-any.whl (1.2 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 884.3/884.3 kB 12.8 MB/s eta 0:00:00
Using cached pycparser-2.22-py3-none-any.whl (117 kB)


> Or choose a different bracket—`bigquery`, `redshift`, etc.—if you prefer another primary destination. For this assignment, we’ll still do a quick test with DuckDB.

2. **Check** the version:


In [2]:
!dlt --version

dlt 1.6.1


or:

In [12]:
import dlt
print("dlt version:", dlt.__version__)

dlt version: 1.6.1


**Answer**:  
- Provide the **version** you see in the output. 1.6.1

## **Question 2: Define & Run the Pipeline (NYC Taxi API)**

Use dlt to extract all pages of data from the API.

Steps:

1️⃣ Use the `@dlt.resource` decorator to define the API source.

2️⃣ Implement automatic pagination using dlt's built-in REST client.

3️⃣ Load the extracted data into DuckDB for querying.



In [4]:
import dlt
from dlt.sources.helpers.rest_client import RESTClient
from dlt.sources.helpers.rest_client.paginators import PageNumberPaginator

@dlt.resource(name="rides", write_disposition="replace")
def ny_taxi():
    client = RESTClient(
        base_url="https://us-central1-dlthub-analytics.cloudfunctions.net",
        paginator=PageNumberPaginator(
            base_page=1,   # Start from page 1
            total_path=None    # API doesn't provide total pages, stop when empty page
        )
    )

    for page in client.paginate("data_engineering_zoomcamp_api"):
        yield page   # Stream data in memory-efficient way

pipeline = dlt.pipeline(
    pipeline_name="ny_taxi_pipeline",
    destination="duckdb",
    dataset_name="ny_taxi_data"
)

Load the data into DuckDB to test:






In [5]:
load_info = pipeline.run(ny_taxi)
print(load_info)

Pipeline ny_taxi_pipeline load step completed in 1.28 seconds
1 load package(s) were loaded to destination duckdb and into dataset ny_taxi_data
The duckdb destination used duckdb:////home/dima/data-engineering-zoomcamp/cohorts/2025/workshops/dlt/ny_taxi_pipeline.duckdb location to store data
Load package 1739136153.653259 is LOADED and contains no failed jobs


Start a connection to your database using native `duckdb` connection and look what tables were generated:

In [9]:
import duckdb

# Connect to the DuckDB database
conn = duckdb.connect(f"{pipeline.pipeline_name}.duckdb")

# Set search path to the dataset
conn.sql(f"SET search_path = '{pipeline.dataset_name}'")

# Describe the dataset
print(conn.sql("DESCRIBE").df())

           database        schema                 name  \
0  ny_taxi_pipeline  ny_taxi_data           _dlt_loads   
1  ny_taxi_pipeline  ny_taxi_data  _dlt_pipeline_state   
2  ny_taxi_pipeline  ny_taxi_data         _dlt_version   
3  ny_taxi_pipeline  ny_taxi_data                rides   

                                        column_names  \
0  [load_id, schema_name, status, inserted_at, sc...   
1  [version, engine_version, pipeline_name, state...   
2  [version, engine_version, inserted_at, schema_...   
3  [end_lat, end_lon, fare_amt, passenger_count, ...   

                                        column_types  temporary  
0  [VARCHAR, VARCHAR, BIGINT, TIMESTAMP WITH TIME...      False  
1  [BIGINT, BIGINT, VARCHAR, VARCHAR, TIMESTAMP W...      False  
2  [BIGINT, BIGINT, TIMESTAMP WITH TIME ZONE, VAR...      False  
3  [DOUBLE, DOUBLE, DOUBLE, BIGINT, VARCHAR, DOUB...      False  


**Answer:**
* How many tables were created? 4

## **Question 3: Explore the loaded data**

Inspect the table `ride`:


In [10]:
df = pipeline.dataset(dataset_type="default").rides.df()
df

end_lat    end_lon  fare_amt  passenger_count payment_type  start_lat  \
0     40.742963 -73.980072      45.0                1       Credit  40.641525   
1     40.740187 -74.005698       6.5                1       Credit  40.722065   
2     40.718043 -74.004745      12.5                5       Credit  40.761945   
3     40.739637 -73.985233       4.9                1         CASH  40.749802   
4     40.730032 -73.852693      25.7                1         CASH  40.776825   
...         ...        ...       ...              ...          ...        ...   
9995  40.783522 -73.970690       5.7                1         CASH  40.778560   
9996  40.777200 -73.964197       4.1                1         CASH  40.779800   
9997  40.780172 -73.957617       6.1                1         CASH  40.788388   
9998  40.777342 -73.957242       5.7                1         CASH  40.773828   
9999  40.757122 -73.986293       6.5                1       Credit  40.756313   

      start_lon  tip_amt  tolls_amt  total_amt  trip_distance  \
0    -73.787442      9.0       4.15      58.15          17.52   
1    -74.009767      1.0       0.00       8.50           1.56   
2    -73.983038      2.0       0.00      15.50           3.37   
3    -73.992247      0.0       0.00       5.40           1.11   
4    -73.949233      0.0       4.15      29.85          11.09   
...         ...      ...        ...        ...            ...   
9995 -73.953660      0.0       0.00       5.70           1.16   
9996 -73.974297      0.0       0.00       4.10           0.89   
9997 -73.976758      0.0       0.00       6.10           1.30   
9998 -73.956690      0.0       0.00       6.20           0.97   
9999 -73.972948      2.0       0.00       8.50           0.92   

        trip_dropoff_date_time     trip_pickup_date_time  surcharge  \
0    2009-06-14 23:48:00+00:00 2009-06-14 23:23:00+00:00        0.0   
1    2009-06-18 17:43:00+00:00 2009-06-18 17:35:00+00:00        1.0   
2    2009-06-10 18:27:00+00:00 2009-06-10 18:08:00+00:00        1.0   
3    2009-06-14 23:58:00+00:00 2009-06-14 23:54:00+00:00        0.5   
4    2009-06-13 13:23:00+00:00 2009-06-13 13:01:00+00:00        0.0   
...                        ...                       ...        ...   
9995 2009-06-19 11:28:00+00:00 2009-06-19 11:22:00+00:00        0.0   
9996 2009-06-17 07:43:00+00:00 2009-06-17 07:41:00+00:00        0.0   
9997 2009-06-19 11:46:00+00:00 2009-06-19 11:39:00+00:00        0.0   
9998 2009-06-17 04:19:00+00:00 2009-06-17 04:13:00+00:00        0.5   
9999 2009-06-17 08:34:00+00:00 2009-06-17 08:24:00+00:00        0.0   

     vendor_name       _dlt_load_id         _dlt_id  store_and_forward  
0            VTS  1739136153.653259  QbWLCQ+azSh9TA                NaN  
1            VTS  1739136153.653259  9zYBaAFWGUyP2A                NaN  
2            VTS  1739136153.653259  A31LKWx/CakJFg                NaN  
3            VTS  1739136153.653259  ESNbIuYv9BZA/g                NaN  
4            VTS  1739136153.653259  j0dvjEbL9Thg+Q                NaN  
...          ...                ...             ...                ...  
9995         VTS  1739136153.653259  agPlaLi/ZkdTCg                NaN  
9996         VTS  1739136153.653259  b9WdrgWmi7d2Hw                NaN  
9997         VTS  1739136153.653259  hXpaEaqFJXyIkQ                NaN  
9998         VTS  1739136153.653259  G/ECbsMTJ+mm9w                NaN  
9999         VTS  1739136153.653259  Bf1xzYZzrlgOIQ                NaN  

[10000 rows x 18 columns]

**Answer:**
* What is the total number of records extracted? 10000

## **Question 4: Trip Duration Analysis**

Run the SQL query below to:

* Calculate the average trip duration in minutes.

In [11]:
with pipeline.sql_client() as client:
    res = client.execute_sql(
            """
            SELECT
            AVG(date_diff('minute', trip_pickup_date_time, trip_dropoff_date_time))
            FROM rides;
            """
        )
    # Prints column values of the first row
    print(res)

[(12.3049,)]


**Answer:**
* What is the average trip duration? 12.3049

## **Submitting the solutions**

* Form for submitting: TBA




## **Solution**

We will publish the solution here after deadline.